File: ui/src/main/java/ui/intro/IntroController.java
package ui.intro;

import com.jfoenix.controls.JFXComboBox;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.scene.layout.VBox;
import model.player.Player;
import model.player.PlayerComputer;
import model.util.Colour;
import ui.game.HumanPlayer;

import java.util.EnumMap;
import java.util.function.Consumer;

/**
 * Controls the main menu window
 */
public class IntroController {
    /**
     * The options for the two players
     */
    private static final ObservableList<Player> PLAYER_OPTIONS = FXCollections.observableArrayList(
            new HumanPlayer(),
            new PlayerComputer(PlayerComputer.EASY),
            new PlayerComputer(PlayerComputer.HARD)
    );

    private final Consumer<EnumMap<Colour, Player>> onStart;

    @FXML
    private VBox whitePlayerContainer;

    @FXML
    private VBox blackPlayerContainer;

    //The drop-downs to select which player is playing
    private final JFXComboBox<Player> whitePlayers = new JFXComboBox<>(PLAYER_OPTIONS);
    private final JFXComboBox<Player> blackPlayers = new JFXComboBox<>(PLAYER_OPTIONS);

    public IntroController(Consumer<EnumMap<Colour, Player>> onStart) {
        this.onStart = onStart;

        //Select the first option
        blackPlayers.getSelectionModel().select(0);
        whitePlayers.getSelectionModel().select(0);
    }

    @FXML
    private void initialize() {
        //Add the drop-downs to the interface
        whitePlayerContainer.getChildren().add(whitePlayers);
        blackPlayerContainer.getChildren().add(blackPlayers);
    }

    /**
     * Called when the start button is pressed
     */
    @FXML
    private void handleStart() {
        //Get the 2 selected players and add them to a EnumMap
        EnumMap<Colour, Player> players = new EnumMap<>(Colour.class);
        players.put(Colour.WHITE, whitePlayers.getSelectionModel().getSelectedItem());
        players.put(Colour.BLACK, blackPlayers.getSelectionModel().getSelectedItem());

        //Call the callback with the players
        onStart.accept(players);
    }
}

File: ui/src/main/java/ui/intro/IntroController.kt
package ui.intro

import com.jfoenix.controls.JFXComboBox
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.layout.VBox
import engine.player.Player
import engine.player.PlayerComputer
import engine.util.Colour
import ui.game.HumanPlayer
import java.util.*

/**
 * Controls the main menu window
 */
class IntroController(private val onStart: (EnumMap<Colour, Player>) -> Unit) {

    @FXML
    private lateinit var whitePlayerContainer: VBox

    @FXML
    private lateinit var blackPlayerContainer: VBox

    //The drop-downs to select which player is playing
    private val whitePlayers = JFXComboBox(PLAYER_OPTIONS)
    private val blackPlayers = JFXComboBox(PLAYER_OPTIONS)

    init {

        //Select the first option
        blackPlayers.selectionModel.select(0)
        whitePlayers.selectionModel.select(0)
    }

    @FXML
    private fun initialize() {
        //Add the drop-downs to the interface
        whitePlayerContainer.children.add(whitePlayers)
        blackPlayerContainer.children.add(blackPlayers)
    }

    /**
     * Called when the start button is pressed
     */
    @FXML
    private fun handleStart() {
        //Get the 2 selected players and add them to a EnumMap
        val players = EnumMap<Colour, Player>(Colour::class.java)
        players[Colour.WHITE] = whitePlayers.selectionModel.selectedItem
        players[Colour.BLACK] = blackPlayers.selectionModel.selectedItem

        //Call the callback with the players
        onStart(players)
    }

    companion object {
        /**
         * The options for the two players
         */
        private val PLAYER_OPTIONS = FXCollections.observableArrayList(
                HumanPlayer(),
                PlayerComputer(PlayerComputer.EASY),
                PlayerComputer(PlayerComputer.HARD)
        )
    }
}